### ISSUE TITLE
Functional attention core: transpose(key) reverses all axes → masked/multi-head paths raise on all inputs

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5648]
## Summary

The functional attention core `src/odyssey/core/attention.mojo` transposes `key` with `transpose(key)` (no `axes` argument), which **reverses ALL axes** instead of swapping only the last two. As a result, the QKᵀ matmul dimension check fails and **every** call to `scaled_dot_product_attention_masked` (3D path) and `multi_head_attention_masked` (4D / multi-head path) raises. The functions are non-executable for their documented input shapes.

## Repro (run under `pixi run mojo -I src -I .`)

**3D** `scaled_dot_product_attention_masked` with q=k=v of shape `[2, 3, 4]` (batch=2, seq=3, d_k=4):

```
Incompatible dimensions for matmul: 4 != 3
```

`transpose(key)` maps `[B, S, d_k] = [2, 3, 4]` → `[4, 3, 2]` (full reversal). The QKᵀ then computes `matmul([2,3,4], [4,3,2])`, whose inner dims `4 (a_cols) != 3 (b_rows)` mismatch → raise.

**4D** `multi_head_attention_masked` with q=k=v `[2, 3, 4]`, `num_heads=2` (d_k=2) and `num_heads=4` (d_k=1):

```
num_heads=2 RAISED: Incompatible dimensions for matmul: 2 != 3
num_heads=4 RAISED: Incompatible dimensions for matmul: 1 != 4
```

`_reshape_for_heads` makes `[B, H, S, d_k]`; `transpose(k_heads)` reverses to `[d_k, S, H, B]` instead of `[B, H, d_k, S]`, so the batched matmul mismatches on every multi-head config.

## Root cause + fix

Two call sites:
- `scaled_dot_product_attention_masked`: `var key_t = transpose(key)` (3D or 4D `key`).
- `multi_head_attention_masked`: `var k_heads_t = transpose(k_heads)` (4D `k_heads`).

Both must swap only the **last two** axes. For an N-D tensor build the identity permutation and swap the final pair, e.g. for 4D pass `axes=[0, 1, 3, 2]`, for 3D `axes=[0, 2, 1]` (`transpose` supports an explicit `axes: List[Int]` permutation). The `*_backward` companions likely need the same audit.

## Reference implementation

The `Module`-conforming `MultiHeadAttention` layer added in PR #5640 (`src/odyssey/core/layers/attention.mojo`) implements the correct last-two-axes transpose and is **parity-verified against PyTorch to 1e-5** (single-head non-causal + multi-head causal). It can serve as the reference when fixing this core, after which the layer's `forward` should delegate to the fixed functional core to remove the duplicate implementation (see PR #5640 review round 1, MAJOR-2).

## Impact

Any current caller of the functional attention API gets a hard raise. PR #5640's layer does not depend on this core, so it is unaffected; the layer carries a `raises`-regression tripwire test that will flag when this bug is fixed so the twin can be collapsed.
